src/ashell/ide-cmd.c

```diff
@@ -0,0 +1,741 @@
+                first = false;
+            } else {
+                ide_spool(",\r\n\t{ \"name\": \"%s\", \"size\": %d }",
+                    entry.name, entry.size);
+            }
+        }
+    };
+    ide_spool("]");  // end of data
+    ide_end_message(NO_ERROR);
+    fs_closedir(&dir);
+    parser_reset();
+}
+
+static void ide_cmd_cat(char *buf, size_t len)
+{
+    IDE_DBG("\r\nInvoking cat...");
+    char *filename = buf;
+    fs_file_t *file;
+    int count, ret;
+
+    if ((ret = parse_filename(buf, len, 1)) > 0) {
+        skip_size(&buf, &len, ret);
+        IDE_DBG("\r\nFilename: %s\r\n", filename);
+
+        if (!fs_exist(filename) || !(file = fs_open_alloc(filename, "r"))) {
+            ide_reply(ERROR_FILE_OPEN, "\"Cannot open file.\"");
+            return;
+        }
+
+        if (fs_size(file) == 0) {
+            fs_close_alloc(file);
+            ide_reply(ERROR_FILE, "\"Empty file.\"");
+            return;
+        }
+
+        // send a separate message for stream start and stream end
+        ide_start_message("cat");
+        ide_spool("\"start\"");  // data
+        ide_end_message(NO_ERROR);
+
+        fs_seek(file, 0, SEEK_SET);
+        do {
+            count = fs_read(file, ide_spool_ptr(), ide_spool_space());
+            ide_spool_adjust(count);
+            ide_spool_flush();
+        } while (count > 0);
+        fs_close_alloc(file);
+
+        ide_reply(NO_ERROR, "\"end\"");
+        return;
+    }
+}
+
+
+// Rename the given file and send back the status.
+static void ide_cmd_move(char *buf, size_t len)
+{
+    parser.state = PARSE_ARG_FILENAME;
+    char *source, *dest;
+
+    IDE_DBG("\r\nInvoking move...");
+    source = buf;
+    int size = parse_filename(buf, len, 1);  // parse as first expected arg
+    if (size < 0)
+        return;  // error has already been handled and parser is reset
+    skip_size(&buf, &len, size);
+    IDE_DBG("\r\nSource: %s", source);
+
+    dest = buf;
+    size = parse_filename(buf, len, 2);  // parse as second expected arg
+    if (size <= 0)
+        return;  // error has already been handled and parser is reset
+    skip_size(&buf, &len, size);
+    IDE_DBG("\r\nDest: %s", dest);
+
+    if (!fs_exist(source)) {
+        ide_reply(ERROR_FILE, "\"Source file not found.\"");
+        return;
+    }
+
+    if (fs_exist(dest)) {
+        ide_reply(ERROR_FILE, "\"Destination already exists.\"");
+        return;
+    }
+
+    f_rename(source, dest);
+    ide_reply(NO_ERROR, "ok");
+}
+
+// Remove the given file and send back the status.
+static void ide_cmd_remove(char *buf, size_t len)
+{
+    char *name = buf;
+
+    IDE_DBG("\r\nInvoking remove...");
+
+    int size = parse_filename(buf, len, 1);
+    if (size <= 0)
+        return;  // error has already been handled and parser is reset
+    skip_size(&buf, &len, size);
+    IDE_DBG("File: %s", name);
+
+    if (!fs_exist(name)) {
+        ide_reply(ERROR_FILE, "\"File not found.\"");
+        return;
+    }
```

Comment:
I'll take it out. Since now the IDE doesn't let users use the console, in theory the chance to try and delete a non-existent file is basically nil. Not to mention yes the unlink should just return an error.